### PR TITLE
catch numpy arrays in attrs before converting to dict

### PR DESCRIPTION
**Pull request by jsignell**
Makes it easier to dump to json (after conversation on #917)

### CODE REVIEW

File: xarray/core/dataarray.py [hunk 1]
```diff
@@ -1194,13 +1194,22 @@ def to_dict(self):
         Convert this xarray.DataArray into a dictionary following xarray
         naming conventions.
 
-        Useful for coverting to json.
+        Useful for coverting to json.To avoid datetime incompatibility
```
Comment (shoyer) [inline xarray/core/dataarray.py]: Missing a space after `json`.

Maybe add "Converts all variables and attributes to native Python objects."

File: xarray/core/dataarray.py [hunk 2]
```diff
@@ -1194,13 +1194,22 @@ def to_dict(self):
...
         --------
         xarray.DataArray.from_dict
         """
-        d = {'coords': {}, 'attrs': dict(self.attrs), 'dims': self.dims}
+        def np_check(attrs):
+            # catch numpy arrays in attrs and convert to nested lists
+            attrs = dict(attrs)
+            for k, v in attrs.items():
+                if hasattr(v, 'tolist'):
+                    attrs.update({k: v.tolist()})
```
Comment (shoyer) [inline xarray/core/dataarray.py]: just use `attrs[k] = v`

File: xarray/core/dataarray.py [hunk 3]
```diff
@@ -1194,13 +1194,22 @@ def to_dict(self):
...
         See also
         --------
         xarray.DataArray.from_dict
         """
-        d = {'coords': {}, 'attrs': dict(self.attrs), 'dims': self.dims}
+        def np_check(attrs):
+            # catch numpy arrays in attrs and convert to nested lists
+            attrs = dict(attrs)
+            for k, v in attrs.items():
+                if hasattr(v, 'tolist'):
```
Comment (shoyer) [inline xarray/core/dataarray.py]: Let's also fix NumPy scalars values, e.g., `np.float32(0)`. To convert these, use `.item()`.

**Comment by jsignell** [comment]
Are the names `time_check` and `np_check` specific enough for utils.py?

**Comment by shoyer** [comment]
Something more descriptive would probably be a good idea. Maybe `ensure_us_time_resolution` and `decode_numpy_dict_values`?

**Comment by shoyer** [comment]
Let's add a test case including a numpy scalar in an attribute? Otherwise, looks good to me.

**Comment by shoyer** [comment]
Thanks!

**Pull request merged**